File: cli/src/llm.ts
import process from 'node:process';
import type { AgentConfig } from './config.js';

export type ChatMessage = {
  role: 'system' | 'user' | 'assistant';
  content: string;
};

export type TokenUsage = {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
};

const EMPTY_USAGE: TokenUsage = {
  promptTokens: 0,
  completionTokens: 0,
  totalTokens: 0,
};

type OpenAiUsage = {
  prompt_tokens?: number;
  completion_tokens?: number;
  total_tokens?: number;
};

type OpenAiResponse = {
  choices?: Array<{
    message?: {
      content?: string;
    };
  }>;
  usage?: OpenAiUsage;
  error?: {
    message?: string;
  };
};

type AnthropicResponse = {
  content?: Array<{
    type?: string;
    text?: string;
  }>;
  usage?: {
    input_tokens?: number;
    output_tokens?: number;
  };
  error?: {
    message?: string;
  };
};

export async function chat(
  cfg: AgentConfig,
  model: string,
  messages: ChatMessage[],
): Promise<{ text: string; usage: TokenUsage }> {
  const provider = cfg.provider.trim().toLowerCase();
  if (provider === 'anthropic') {
    return chatAnthropic(cfg, model, messages);
  }

  return chatOpenAi(cfg, model, messages);
}

async function chatOpenAi(
  cfg: AgentConfig,
  model: string,
  messages: ChatMessage[],
): Promise<{ text: string; usage: TokenUsage }> {
  if (!cfg.baseUrl.trim()) {
    throw new Error('llm base_url is required');
  }
  if (!model.trim()) {
    throw new Error('llm model is required');
  }

  const url = `${cfg.baseUrl.replace(/\/+$/, '')}${cfg.chatCompletionsPath}`;
  const body = {
    model,
    messages,
    temperature: cfg.temperature,
    max_tokens: cfg.maxTokens,
  };

  const headers = resolveHeaders(cfg);
  if (!hasHeader(headers, 'content-type')) {
    headers['Content-Type'] = 'application/json';
  }
  if (!hasHeader(headers, 'user-agent')) {
    headers['User-Agent'] = 'glossia';
  }

  const timeout = (cfg.timeoutSeconds > 0 ? cfg.timeoutSeconds : 300) * 1000;
  const response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(timeout),
  });

  const rawBody = await response.text();
  const normalized = normalizeResponseBody(rawBody);
  const parsed = JSON.parse(normalized) as OpenAiResponse;

  if (response.status >= 400) {
    throw new Error(`llm error: ${parsed.error?.message ?? `status ${response.status}`}`);
  }

  const firstChoice = parsed.choices?.[0];
  const text = firstChoice?.message?.content ?? '';
  if (!text) {
    throw new Error('llm response missing choices');
  }

  return {
    text,
    usage: {
      promptTokens: parsed.usage?.prompt_tokens ?? 0,
      completionTokens: parsed.usage?.completion_tokens ?? 0,
      totalTokens: parsed.usage?.total_tokens ?? 0,
    },
  };
}

async function chatAnthropic(
  cfg: AgentConfig,
  model: string,
  messages: ChatMessage[],
): Promise<{ text: string; usage: TokenUsage }> {
  if (!cfg.baseUrl.trim()) {
    throw new Error('llm base_url is required');
  }
  if (!model.trim()) {
    throw new Error('llm model is required');
  }

  const url = `${cfg.baseUrl.replace(/\/+$/, '')}${cfg.chatCompletionsPath}`;

  const systemParts: string[] = [];
  const anthropicMessages: Array<{ role: 'user' | 'assistant'; content: string }> = [];

  for (const message of messages) {
    if (message.role === 'system') {
      if (message.content.trim()) {
        systemParts.push(message.content);
      }
      continue;
    }

    if (message.role === 'user' || message.role === 'assistant') {
      anthropicMessages.push({
        role: message.role,
        content: message.content,
      });
    }
  }

  if (anthropicMessages.length === 0) {
    throw new Error('llm request requires user messages');
  }

  const headers = resolveHeaders(cfg);
  if (!hasHeader(headers, 'content-type')) {
    headers['Content-Type'] = 'application/json';
  }
  if (!hasHeader(headers, 'user-agent')) {
    headers['User-Agent'] = 'glossia';
  }

  const timeout = (cfg.timeoutSeconds > 0 ? cfg.timeoutSeconds : 300) * 1000;
  const response = await fetch(url, {
    method: 'POST',
    headers,
    signal: AbortSignal.timeout(timeout),
    body: JSON.stringify({
      model,
      max_tokens: cfg.maxTokens && cfg.maxTokens > 0 ? cfg.maxTokens : 1024,
      messages: anthropicMessages,
      system: systemParts.length > 0 ? systemParts.join('\n\n') : undefined,
      temperature: cfg.temperature,
    }),
  });

  const parsed = (await response.json()) as AnthropicResponse;
  if (response.status >= 400) {
    throw new Error(`llm error: ${parsed.error?.message ?? `status ${response.status}`}`);
  }

  const text =
    parsed.content
      ?.filter((block) => block.type === 'text')
      .map((block) => block.text ?? '')
      .join('') ?? '';

  if (!text) {
    throw new Error('llm response missing text');
  }

  const input = parsed.usage?.input_tokens ?? 0;
  const output = parsed.usage?.output_tokens ?? 0;

  return {
    text,
    usage: {
      promptTokens: input,
      completionTokens: output,
      totalTokens: input + output,
    },
  };
}

function normalizeResponseBody(body: string): string {
  const trimmed = body.trim();

  if (!trimmed.startsWith('[')) {
    return trimmed;
  }

  const parsed = JSON.parse(trimmed) as unknown[];
  if (parsed.length !== 1) {
    throw new Error(`llm response: unexpected array with ${parsed.length} elements`);
  }

  const first = parsed[0] as { error?: { message?: string } };
  if (first?.error?.message) {
    throw new Error(`llm error: ${first.error.message}`);
  }

  return JSON.stringify(first);
}

function resolveHeaders(cfg: AgentConfig): Record<string, string> {
  const headers: Record<string, string> = {};

  for (const [key, value] of Object.entries(cfg.headers)) {
    headers[key] = expandEnv(value);
  }

  const provider = cfg.provider.trim().toLowerCase();
  if (provider === 'anthropic') {
    if (!hasHeader(headers, 'x-api-key')) {
      const key = resolveApiKey(cfg);
      if (key) {
        headers['x-api-key'] = key;
      }
    }
    if (!hasHeader(headers, 'anthropic-version')) {
      headers['anthropic-version'] = '2023-06-01';
    }
    return headers;
  }

  if (!hasHeader(headers, 'authorization')) {
    const key = resolveApiKey(cfg);
    if (key) {
      headers.Authorization = `Bearer ${key}`;
    }
  }

  return headers;
}

function resolveApiKey(cfg: AgentConfig): string {
  const fromInline = expandEnv(cfg.apiKey).trim();
  if (fromInline) {
    return fromInline;
  }

  if (!cfg.apiKeyEnv.trim()) {
    return '';
  }

  return process.env[cfg.apiKeyEnv] ?? '';
}

function hasHeader(headers: Record<string, string>, name: string): boolean {
  const normalized = name.toLowerCase();
  return Object.keys(headers).some((header) => header.toLowerCase() === normalized);
}

function expandEnv(input: string): string {
  const withTemplate = input.replace(/\{\{\s*env\.([A-Za-z_][A-Za-z0-9_]*)\s*\}\}/g, (_, key) => {
    return process.env[key] ?? '';
  });

  if (withTemplate.startsWith('env.')) {
    return process.env[withTemplate.slice(4)] ?? '';
  }

  return withTemplate.replace(/env:([A-Za-z_][A-Za-z0-9_]*)/g, (_, key) => process.env[key] ?? '');
}

export function addUsage(sum: TokenUsage, next: TokenUsage): TokenUsage {
  return {
    promptTokens: sum.promptTokens + next.promptTokens,
    completionTokens: sum.completionTokens + next.completionTokens,
    totalTokens: sum.totalTokens + next.totalTokens,
  };
}

export function emptyUsage(): TokenUsage {
  return { ...EMPTY_USAGE };
}

File: cli/internal/glossia/llm.go
package glossia

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

type ChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type TokenUsage struct {
	PromptTokens     int
	CompletionTokens int
	TotalTokens      int
}

var emptyUsageValue = TokenUsage{PromptTokens: 0, CompletionTokens: 0, TotalTokens: 0}

func emptyUsage() TokenUsage { return emptyUsageValue }

func addUsage(sum TokenUsage, next TokenUsage) TokenUsage {
	return TokenUsage{
		PromptTokens:     sum.PromptTokens + next.PromptTokens,
		CompletionTokens: sum.CompletionTokens + next.CompletionTokens,
		TotalTokens:      sum.TotalTokens + next.TotalTokens,
	}
}

func chat(cfg AgentConfig, model string, messages []ChatMessage) (string, TokenUsage, error) {
	provider := strings.ToLower(strings.TrimSpace(cfg.Provider))
	if provider == "anthropic" {
		return chatAnthropic(cfg, model, messages)
	}
	return chatOpenAI(cfg, model, messages)
}

type openAIResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
		TotalTokens      int `json:"total_tokens"`
	} `json:"usage"`
	Error *struct {
		Message string `json:"message"`
	} `json:"error"`
}

func chatOpenAI(cfg AgentConfig, model string, messages []ChatMessage) (string, TokenUsage, error) {
	if strings.TrimSpace(cfg.BaseURL) == "" {
		return "", TokenUsage{}, fmt.Errorf("llm base_url is required")
	}
	if strings.TrimSpace(model) == "" {
		return "", TokenUsage{}, fmt.Errorf("llm model is required")
	}

	url := strings.TrimRight(cfg.BaseURL, "/") + cfg.ChatCompletionsPath

	body := map[string]any{
		"model":    model,
		"messages": messages,
	}
	if cfg.Temperature != nil {
		body["temperature"] = *cfg.Temperature
	}
	if cfg.MaxTokens != nil {
		body["max_tokens"] = *cfg.MaxTokens
	}

	headers := resolveHeaders(cfg)
	if !hasHeader(headers, "content-type") {
		headers["Content-Type"] = "application/json"
	}
	if !hasHeader(headers, "user-agent") {
		headers["User-Agent"] = "glossia"
	}

	timeout := 300 * time.Second
	if cfg.TimeoutSeconds > 0 {
		timeout = time.Duration(cfg.TimeoutSeconds) * time.Second
	}

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	encoded, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(encoded))
	if err != nil {
		return "", TokenUsage{}, err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", TokenUsage{}, err
	}
	defer resp.Body.Close()

	rawBody, _ := io.ReadAll(resp.Body)
	normalized, err := normalizeResponseBody(rawBody)
	if err != nil {
		return "", TokenUsage{}, err
	}

	var parsed openAIResponse
	if err := json.Unmarshal(normalized, &parsed); err != nil {
		return "", TokenUsage{}, err
	}

	if resp.StatusCode >= 400 {
		msg := fmt.Sprintf("status %d", resp.StatusCode)
		if parsed.Error != nil && strings.TrimSpace(parsed.Error.Message) != "" {
			msg = parsed.Error.Message
		}
		return "", TokenUsage{}, fmt.Errorf("llm error: %s", msg)
	}

	text := ""
	if len(parsed.Choices) > 0 {
		text = parsed.Choices[0].Message.Content
	}
	if strings.TrimSpace(text) == "" {
		return "", TokenUsage{}, fmt.Errorf("llm response missing choices")
	}

	return text, TokenUsage{
		PromptTokens:     parsed.Usage.PromptTokens,
		CompletionTokens: parsed.Usage.CompletionTokens,
		TotalTokens:      parsed.Usage.TotalTokens,
	}, nil
}

type anthropicResponse struct {
	Content []struct {
		Type string `json:"type"`
		Text string `json:"text"`
	} `json:"content"`
	Usage struct {
		InputTokens  int `json:"input_tokens"`
		OutputTokens int `json:"output_tokens"`
	} `json:"usage"`
	Error *struct {
		Message string `json:"message"`
	} `json:"error"`
}

func chatAnthropic(cfg AgentConfig, model string, messages []ChatMessage) (string, TokenUsage, error) {
	if strings.TrimSpace(cfg.BaseURL) == "" {
		return "", TokenUsage{}, fmt.Errorf("llm base_url is required")
	}
	if strings.TrimSpace(model) == "" {
		return "", TokenUsage{}, fmt.Errorf("llm model is required")
	}

	url := strings.TrimRight(cfg.BaseURL, "/") + cfg.ChatCompletionsPath

	var systemParts []string
	var anthropicMessages []map[string]string

	for _, m := range messages {
		if m.Role == "system" {
			if strings.TrimSpace(m.Content) != "" {
				systemParts = append(systemParts, m.Content)
			}
			continue
		}

		if m.Role == "user" || m.Role == "assistant" {
			anthropicMessages = append(anthropicMessages, map[string]string{
				"role":    m.Role,
				"content": m.Content,
			})
		}
	}

	if len(anthropicMessages) == 0 {
		return "", TokenUsage{}, fmt.Errorf("llm request requires user messages")
	}

	maxTokens := 1024
	if cfg.MaxTokens != nil && *cfg.MaxTokens > 0 {
		maxTokens = *cfg.MaxTokens
	}

	body := map[string]any{
		"model":      model,
		"max_tokens": maxTokens,
		"messages":   anthropicMessages,
	}
	if cfg.Temperature != nil {
		body["temperature"] = *cfg.Temperature
	}
	if len(systemParts) > 0 {
		body["system"] = strings.Join(systemParts, "\n\n")
	}

	headers := resolveHeaders(cfg)
	if !hasHeader(headers, "content-type") {
		headers["Content-Type"] = "application/json"
	}
	if !hasHeader(headers, "user-agent") {
		headers["User-Agent"] = "glossia"
	}

	timeout := 300 * time.Second
	if cfg.TimeoutSeconds > 0 {
		timeout = time.Duration(cfg.TimeoutSeconds) * time.Second
	}

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	encoded, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(encoded))
	if err != nil {
		return "", TokenUsage{}, err
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", TokenUsage{}, err
	}
	defer resp.Body.Close()

	rawBody, _ := io.ReadAll(resp.Body)

	var parsed anthropicResponse
	if err := json.Unmarshal(rawBody, &parsed); err != nil {
		return "", TokenUsage{}, err
	}

	if resp.StatusCode >= 400 {
		msg := fmt.Sprintf("status %d", resp.StatusCode)
		if parsed.Error != nil && strings.TrimSpace(parsed.Error.Message) != "" {
			msg = parsed.Error.Message
		}
		return "", TokenUsage{}, fmt.Errorf("llm error: %s", msg)
	}

	var text strings.Builder
	for _, block := range parsed.Content {
		if block.Type == "text" {
			text.WriteString(block.Text)
		}
	}

	outText := text.String()
	if strings.TrimSpace(outText) == "" {
		return "", TokenUsage{}, fmt.Errorf("llm response missing text")
	}

	input := parsed.Usage.InputTokens
	output := parsed.Usage.OutputTokens

	return outText, TokenUsage{
		PromptTokens:     input,
		CompletionTokens: output,
		TotalTokens:      input + output,
	}, nil
}

func normalizeResponseBody(body []byte) ([]byte, error) {
	trimmed := bytes.TrimSpace(body)
	if len(trimmed) == 0 {
		return trimmed, nil
	}
	if trimmed[0] != '[' {
		return trimmed, nil
	}

	var arr []json.RawMessage
	if err := json.Unmarshal(trimmed, &arr); err != nil {
		return nil, err
	}
	if len(arr) != 1 {
		return nil, fmt.Errorf("llm response: unexpected array with %d elements", len(arr))
	}

	var first struct {
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	_ = json.Unmarshal(arr[0], &first)
	if first.Error != nil && strings.TrimSpace(first.Error.Message) != "" {
		return nil, fmt.Errorf("llm error: %s", first.Error.Message)
	}

	return arr[0], nil
}

func resolveHeaders(cfg AgentConfig) map[string]string {
	headers := map[string]string{}
	for k, v := range cfg.Headers {
		headers[k] = expandEnv(v)
	}

	provider := strings.ToLower(strings.TrimSpace(cfg.Provider))
	if provider == "anthropic" {
		if !hasHeader(headers, "x-api-key") {
			key := resolveAPIKey(cfg)
			if strings.TrimSpace(key) != "" {
				headers["x-api-key"] = key
			}
		}
		if !hasHeader(headers, "anthropic-version") {
			headers["anthropic-version"] = "2023-06-01"
		}
		return headers
	}

	if !hasHeader(headers, "authorization") {
		key := resolveAPIKey(cfg)
		if strings.TrimSpace(key) != "" {
			headers["Authorization"] = "Bearer " + key
		}
	}

	return headers
}

func resolveAPIKey(cfg AgentConfig) string {
	fromInline := strings.TrimSpace(expandEnv(cfg.APIKey))
	if fromInline != "" {
		return fromInline
	}

	if strings.TrimSpace(cfg.APIKeyEnv) == "" {
		return ""
	}
	return os.Getenv(cfg.APIKeyEnv)
}

func hasHeader(headers map[string]string, name string) bool {
	normalized := strings.ToLower(name)
	for k := range headers {
		if strings.ToLower(k) == normalized {
			return true
		}
	}
	return false
}

var envTemplateRe = regexp.MustCompile(`\{\{\s*env\.([A-Za-z_][A-Za-z0-9_]*)\s*\}\}`)
var envInlineRe = regexp.MustCompile(`env:([A-Za-z_][A-Za-z0-9_]*)`)

func expandEnv(input string) string {
	withTemplate := envTemplateRe.ReplaceAllStringFunc(input, func(match string) string {
		sub := envTemplateRe.FindStringSubmatch(match)
		if len(sub) != 2 {
			return ""
		}
		return os.Getenv(sub[1])
	})

	if strings.HasPrefix(withTemplate, "env.") {
		return os.Getenv(strings.TrimPrefix(withTemplate, "env."))
	}

	return envInlineRe.ReplaceAllStringFunc(withTemplate, func(match string) string {
		sub := envInlineRe.FindStringSubmatch(match)
		if len(sub) != 2 {
			return ""
		}
		return os.Getenv(sub[1])
	})
}
